### PR TITLE
KOGITO-3279 - more robust LIME perturbation selection

### PR DESCRIPTION
**Pull request by tteofili**
Make LIME choice of the number of features to perturb more robust, especially when working with lists of inputs (hence the no. of features is not known in advance).
This change make perturbation happen not on a fixed number of features but rather on a number of features that is constrained to a `[|features|/2, noOfPerturbations]` range (the range is flipped in case `noOfPerturbations` is smaller than half the number of existing features in the input).
Note that the smallest lower bound for perturbation is kept to `1` to avoid not perturbing _any_ feature.

See https://issues.redhat.com/browse/KOGITO-3279

Many thanks for submitting your Pull Request :heart:! 

Please make sure that your PR meets the following requirements:

- [x] You have read the [contributors guide](https://github.com/kiegroup/kogito-runtimes#contributing-to-kogito)
- [x] Pull Request title is properly formatted: `KOGITO-XYZ Subject`
- [x] Pull Request title contains the target branch if not targeting master: `[0.9.x] KOGITO-XYZ Subject`
- [x] Pull Request contains link to the JIRA issue
- [x] Pull Request contains link to any dependent or related Pull Request
- [x] Pull Request contains description of the issue
- [x] Pull Request does not include fixes for issues other than the main ticket

### CODE REVIEW

File: explainability/explainability-core/src/main/java/org/kie/kogito/explainability/model/Type.java [hunk 1]
```diff
@@ -314,12 +314,26 @@
...
+                else if (upperBound > lowerBound) {
+                    perturbationSize = perturbationContext.getRandom().ints(lowerBound, 1 + upperBound).findFirst().orElse(1);
+                }
+                if (perturbationSize > 0) {
+                    int[] indexesToBePerturbed = perturbationContext.getRandom().ints(1, newList.size())
+                            .distinct().limit(perturbationSize).toArray();
+                    for (int index : indexesToBePerturbed) {
+                        Feature cf = composite.get(index);
+                        Feature f = FeatureFactory.copyOf(cf, cf.getType().perturb(cf.getValue(), perturbationContext));
+                        newList.set(index, f);
```
Comment (danielezonca) [inline explainability/explainability-core/src/main/java/org/kie/kogito/explainability/model/Type.java]: This line seems to me the only difference of this code compared to the other changed method. Can you please refactor to a static utility method?
I expect it should be possible to create a method that accepts `List<Feature> inputFeatures` and `PerturbationContext perturbationContext` and returns `List<Feature>`

Comment (tteofili) [reply explainability/explainability-core/src/main/java/org/kie/kogito/explainability/model/Type.java]: sure, it makes sense to refactor it in a utility method.

**Comment by tteofili** [comment]
jenkins retest this please

**Review by danielezonca — APPROVED** [review]
LGMT

**Comment by tteofili** [comment]
jenkins retest this please

**Comment by danielezonca** [comment]
@tteofili 
One of the previous build failed with
`testPerturbCompositeFeature – org.kie.kogito.explainability.model.TypeTest` ( 
https://rhba-jenkins.rhev-ci-vms.eng.rdu2.redhat.com/blue/organizations/jenkins/KIE%2Fkogito%2Fghprb-webhooks%2Fkogito-apps/detail/kogito-apps/1316/tests )
Now is not failing but I suspect it is a random failure, so can you please double check to be sure the test is stable?

**Comment by tteofili** [comment]
@danielezonca I've found the issue, it's unrelated to the change, but actually a bug for perturbation of `Type.Boolean` `Features`.
The bug is in the wrong usage of `Boolean.getBoolean(value)` while `Boolean.parseBoolean(value)` should be used at https://github.com/kiegroup/kogito-apps/blob/master/explainability/explainability-core/src/main/java/org/kie/kogito/explainability/model/Type.java#L162

**Review by jiripetrlik — APPROVED** [review]
LGTM

**Pull request merged**